src/client/websocket/WebSocketManager.js

```diff
@@ -37,52 +77,157 @@ class WebSocketManager extends EventEmitter {
+   * @param {Object} packet The packet to be handled
+   * @param {WebSocketShard} shard The shard that will handle this packet
    * @returns {boolean}
    */
-  connect(gateway) {
-    if (!this.connection) {
-      this.connection = new WebSocketConnection(this, gateway);
-      return true;
-    }
-    switch (this.connection.status) {
-      case Status.IDLE:
-      case Status.DISCONNECTED:
-        this.connection.connect(gateway, 5500);
-        return true;
-      default:
-        this.debug(`Couldn't connect to ${gateway} as the websocket is at state ${this.connection.status}`);
+  handlePacket(packet, shard) {
+    if (packet && this.status !== Status.READY) {
+      if (!BeforeReadyWhitelist.includes(packet.t)) {
+        this.packetQueue.push({ packet, shardID: shard.id });
         return false;
+      }
+    }
+
+    if (this.packetQueue.length) {
+      const item = this.packetQueue.shift();
+      this.client.setImmediate(() => {
+        this.handlePacket(item.packet, this.shards[item.shardID]);
+      });
+    }
+
+    if (packet && PacketHandlers[packet.t]) {
+      PacketHandlers[packet.t](this.client, packet, shard);
+    }
+
+    return false;
+  }
+
+  /**
+   * Checks whether the client is ready to be marked as ready.
+   * @returns {void}
+   */
+  checkReady() {
+    if (!(this.shards.filter(s => s).length === this.client.options.shardCount) ||
+      !this.shards.every(s => s.status === Status.READY)) {
+      return false;
+    }
+
+    let unavailableGuilds = 0;
+    for (const guild of this.client.guilds.values()) {
+      if (!guild.available) unavailableGuilds++;
+    }
+    if (unavailableGuilds === 0) {
+      this.status = Status.NEARLY;
+      if (!this.client.options.fetchAllMembers) return this.triggerReady();
+      // Fetch all members before marking self as ready
+      const promises = this.client.guilds.map(g => g.members.fetch());
+      Promise.all(promises)
+        .then(() => this.triggerReady())
+        .catch(e => {
+          this.debug(`Failed to fetch all members before ready! ${e}`);
+          this.triggerReady();
+        });
+    }
+    return true;
+  }
+
+  /**
+   * Causes the client to be marked as ready and emits the ready event.
+   * @returns {void}
+   */
+  triggerReady() {
+    if (this.status === Status.READY) {
+      this.debug('Tried to mark self as ready, but already ready');
+      return;
+    }
+    /**
+     * Emitted when the client becomes ready to start working.
+     * @event Client#ready
+     */
+    this.status = Status.READY;
+    this.client.emit(Events.READY);
+    this.handlePacket();
+  }
+
+  /*
+   * Broadcasts a message to every shard in this WebSocketManager.
+   * @param {*} packet The packet to send
+   */
+  broadcast(packet) {
+    for (const shard of this.shards) {
+      shard.send(packet);
+    }
+  }
+
+  destroy() {
+    for (const shard of this.shards) {
+      this.gateway = undefined;
```

Comment:
Just a quick thought, shouldn't this be `shard.gateway = undefined;` and `shard.spawning = true;`?

Reply:
those two lines should have been outside the loop... @hydrabolt please fix